R/post-parsers.R

```diff
@@ -0,0 +1,183 @@
+#' ```r
+#' parser <- function(...) {
+#'   function(value, ...) {
+#'     # do something with raw value
+#'   }
+#' }
+#' ```
+#'
+#' It should return a named list if you want values to map to
+#' plumber endpoint function args.
+#'
+#' @example
+#' parser_json <- function(...) {
+#'   function(value, content_type = "application/json", ...) {
+#'     charset <- getCharacterSet(content_type)
+#'     value <- rawToChar(value)
+#'     Encoding(value) <- charset
+#'     safeFromJSON(value)
+#'   }
+#' }
+#' @md
+#' @export
+addParser <- function(name, parser, pattern = NULL) {
+  if (is.null(.globals$parsers)) {
+    .globals$parsers <- list()
+  }
+  if (!is.null(.globals$parsers$f[[name]])) {
+    stop("Already have a parser by the name of ", name)
+  }
+  if (is.null(pattern)) {
+    pattern <- paste0("application/", name)
+  }
+  .globals$parsers$f[[name]] <- parser
+  .globals$parsers$p[[name]] <- pattern
+}
+
+
+
+#' JSON
+#' @rdname parsers
+#' @param ... Raw values and headers are passed there.
+#' @export
+parser_json <- function(...) {
+  function(value, content_type = NULL, ...) {
+    charset <- getCharacterSet(content_type)
+    value <- rawToChar(value)
+    Encoding(value) <- charset
+    safeFromJSON(value)
+  }
+}
+
+
+
+
+#' QUERY STRING
+#' @rdname parsers
+#' @export
+parser_query <- function(...) {
+  function(value, content_type = NULL, ...) {
+    charset <- getCharacterSet(content_type)
+    value <- rawToChar(value)
+    Encoding(value) <- charset
+    parseQS(value)
+  }
+}
+
+
+
+
+#' TEXT
+#' @rdname parsers
+#' @export
+parser_text <- function(...) {
+  function(value, content_type = NULL, ...) {
+    charset <- getCharacterSet(content_type)
+    value <- rawToChar(value)
+    Encoding(value) <- charset
+    value
+  }
+}
+
+
+
+
+#" RDS
+#' @rdname parsers
+#' @export
+parser_rds <- function(...) {
+  function(value, filename, ...) {
+    tmp <- tempfile("plumb", fileext = paste0("_", basename(filename)))
+    writeBin(value, tmp)
+    on.exit(file.remove(tmp))
+    list(readRDS(tmp))
+  }
+}
+
+
+
+
+#" MULTI
+#' @rdname parsers
+#' @export
+#' @importFrom webutils parse_multipart
+parser_multi <- function(...) {
+  function(value, content_type, ...) {
+    if (!stri_detect_fixed(content_type, "boundary=", case_insensitive = TRUE))
+      stop("No boundary found in multipart content-type header: ", content_type)
+    boundary <- stri_match_first_regex(content_type, "boundary=([^; ]{2,})", case_insensitive = TRUE)[,2]
+    toparse <- parse_multipart(value, boundary)
+    # content-type detection
+    for (i in seq_len(length(toparse))) {
+      if (!is.null(toparse[[i]]$filename)) {
+        ext <- tools::file_ext(toparse[[i]]$filename)
+        toparse[[i]]$content_type <- getContentType(ext)
+      }
+    }
+    lapply(toparse, parseRaw)
+  }
+}
+
+
+
+
+#' OCTET
+#' @rdname parsers
+#' @param ... Raw values and headers are passed there.
+#' @export
+parser_octet <- function(...) {
```

Comment:
return raw by default. Allow for a option / parameter to be set to return the file name (of an uncleared file).

Reply:
Not sure about the option name, any suggestion welcome